test/box-luatest/error_subsystem_improvements_test.lua

```diff
@@ -1050,3 +1051,15 @@ g.test_illegal_params = function()
     t.assert_equals(e.type, 'IllegalParams')
     t.assert_equals(e.message, 'foo')
 end
+
+g.test_luac_error_trace = function()
```

Comment:
It'd be better if you tested `luaT_error` and `luaT_error_at` instead in a unit test.

Reply:
Done.

<details></details>
<summary>diff</summary>

```diff
diff --git a/test/box-luatest/error_subsystem_improvements_test.lua b/test/box-luatest/error_subsystem_improvements_test.lua
index 8a261287d..3ffbba22b 100644
--- a/test/box-luatest/error_subsystem_improvements_test.lua
+++ b/test/box-luatest/error_subsystem_improvements_test.lua
@@ -4,7 +4,6 @@ local console = require('console')
 local json = require('json')
 local yaml = require('yaml')
 local tarantool = require('tarantool')
-local popen = require('popen')
 local server = require('luatest.server')
 
 local g = t.group()
@@ -1051,15 +1050,3 @@ g.test_illegal_params = function()
     t.assert_equals(e.type, 'IllegalParams')
     t.assert_equals(e.message, 'foo')
 end
-
-g.test_luac_error_trace = function()
-    t.assert_equals(debug.getinfo(popen.new, 'S').short_src, '[C]')
-    local this_file = debug.getinfo(1, 'S').short_src
-    local line = debug.getinfo(1, 'l').currentline + 1
-    local wrapper = function() popen.new(1) end
-    local ok, err = pcall(wrapper)
-    t.assert_not(ok)
-    t.assert(box.error.is(err))
-    t.assert_equals(err.trace[1].file, this_file)
-    t.assert_equals(err.trace[1].line, line)
-end
diff --git a/test/unit/CMakeLists.txt b/test/unit/CMakeLists.txt
index 16aa57f15..9a64e732d 100644
--- a/test/unit/CMakeLists.txt
+++ b/test/unit/CMakeLists.txt
@@ -638,6 +638,15 @@ create_unit_test(PREFIX lua_utils
                            ${READLINE_LIBRARIES}
 )
 
+create_unit_test(PREFIX lua_error
+                 SOURCES lua_error.c
+                 LIBRARIES unit core server
+                           ${LUAJIT_LIBRARIES}
+                           ${CURL_LIBRARIES}
+                           ${LIBYAML_LIBRARIES}
+                           ${READLINE_LIBRARIES}
+)
+
 create_unit_test(PREFIX lua_msgpack
                  SOURCES lua_msgpack.c
                  LIBRARIES unit core server
diff --git a/test/unit/lua_error.c b/test/unit/lua_error.c
new file mode 100644
index 000000000..86afedff3
--- /dev/null
+++ b/test/unit/lua_error.c
@@ -0,0 +1,117 @@
+#include <stdio.h>
+
+#include "diag.h"
+#include "fiber.h"
+#include "lualib.h"
+#include "memory.h"
+
+#include "lua/utils.h"
+
+#define UNIT_TAP_COMPATIBLE 1
+#include "unit.h"
+
+static int
+raise_error(lua_State *L)
+{
+	diag_set(IllegalParams, "foo");
+	return luaT_error(L);
+}
+
+static int
+raise_error_at(lua_State *L)
+{
+	int level = lua_tointeger(L, 1);
+	diag_set(IllegalParams, "bar");
+	return luaT_error_at(L, level);
+}
+
+static int
+error_trace(lua_State *L)
+{
+	struct error *e = luaL_checkerror(L, 1);
+
+	lua_createtable(L, 0, 2);
+	lua_pushstring(L, "file");
+	lua_pushstring(L, e->file);
+	lua_settable(L, -3);
+	lua_pushstring(L, "line");
+	lua_pushinteger(L, e->line);
+	lua_settable(L, -3);
+	return 1;
+}
+
+const char *test_error_lua =
+"local this_file = debug.getinfo(1, 'S').short_src\n"
+"local line1 = debug.getinfo(1, 'l').currentline + 1\n"
+"local f1 = function(fn, ...) fn(...) end\n"
+"local line2 = debug.getinfo(1, 'l').currentline + 1\n"
+"local f2 = function(fn, ...) f1(fn, ...) end\n"
+"local line3 = debug.getinfo(1, 'l').currentline + 1\n"
+"local f3 = function(fn, ...) f2(fn, ...) end\n"
+"\n"
+"local function check(line, fn, ...)\n"
+"    local ok, err = pcall(f3, fn, ...)\n"
+"\n"
+"    assert(not ok, string.format('got %s', err))\n"
+"    local trace = test_error_trace(err)\n"
+"    assert(trace.file == this_file, string.format('got \"%s\"', trace.file))\n"
+"    assert(trace.line == line,\n"
+"           string.format('expected %d, got %d', line, trace.line))\n"
+"end\n"
+"\n"
+"assert(test_raise_error ~= nil)"
+"check(line1, test_raise_error)\n"
+"check(line1, test_raise_error_at, 1)\n"
+"check(line2, test_raise_error_at, 2)\n"
+"check(line3, test_raise_error_at, 3)\n"
+"\n"
+"local ok, err = pcall(f3, test_raise_error_at, 0)\n"
+"assert(not ok, string.format('got %s', err))\n"
+"local trace = test_error_trace(err)\n"
+"assert(trace.file == './test/unit/lua_error.c', "
+"       string.format('got %s', trace.file))\n";
+
+static void
+test_error(lua_State *L)
+{
+	plan(1);
+	header();
+
+	int rc;
+
+	lua_pushcfunction(L, raise_error);
+	lua_setglobal(L, "test_raise_error");
+	lua_pushcfunction(L, raise_error_at);
+	lua_setglobal(L, "test_raise_error_at");
+	lua_pushcfunction(L, error_trace);
+	lua_setglobal(L, "test_error_trace");
+
+	rc = luaT_dostring(L, test_error_lua);
+	ok(rc == 0, rc == 0 ? "OK" : "got %s",
+	   diag_last_error(diag_get())->errmsg);
+
+	footer();
+	check_plan();
+}
+
+int
+main(void)
+{
+	plan(1);
+	header();
+
+	struct lua_State *L = luaL_newstate();
+	luaL_openlibs(L);
+	memory_init();
+	fiber_init(fiber_c_invoke);
+	tarantool_lua_error_init(L);
+
+	test_error(L);
+
+	fiber_free();
+	memory_free();
+	lua_close(L);
+
+	footer();
+	return check_plan();
+}

```

</details>